src/builder.rs

```diff
@@ -178,6 +175,284 @@ impl<'ctx> Builder<'ctx> {
+    ///
+    /// This function can take either a `FunctionValue` or a `PointerValue`
+    /// which is a function pointer. It will panic if the `PointerValue` is not a function pointer.
+    /// This may be turned into a Result in the future, however.
+    pub fn build_invoke<F>(
+        &self,
+        function: F,
+        args: &[BasicValueEnum<'ctx>],
+        then_block: BasicBlock<'ctx>,
+        catch_block: BasicBlock<'ctx>,
+        name: &str,
+    ) -> CallSiteValue<'ctx>
+    where
+        F: Into<FunctionOrPointerValue<'ctx>>,
+    {
+        let fn_val_ref = match function.into() {
+            Left(val) => val.as_value_ref(),
+            Right(val) => {
+                // If using a pointer value, we must validate it's a valid function ptr
+                let value_ref = val.as_value_ref();
+                let ty_kind = unsafe { LLVMGetTypeKind(LLVMGetElementType(LLVMTypeOf(value_ref))) };
+                let is_a_fn_ptr = matches!(ty_kind, LLVMTypeKind::LLVMFunctionTypeKind);
+
+                // REVIEW: We should probably turn this into a Result?
+                assert!(
+                    is_a_fn_ptr,
+                    "build_call called with a pointer which is not a function pointer"
+                );
+
+                value_ref
+            }
+        };
+
+        // LLVM gets upset when void return calls are named because they don't return anything
+        let name = unsafe {
+            match LLVMGetTypeKind(LLVMGetReturnType(LLVMGetElementType(LLVMTypeOf(
+                fn_val_ref,
+            )))) {
+                LLVMTypeKind::LLVMVoidTypeKind => "",
+                _ => name,
+            }
+        };
+
+        let c_string = to_c_str(name);
+        let mut args: Vec<LLVMValueRef> = args.iter().map(|val| val.as_value_ref()).collect();
+        let value = unsafe {
+            LLVMBuildInvoke(
+                self.builder,
+                fn_val_ref,
+                args.as_mut_ptr(),
+                args.len() as u32,
+                then_block.basic_block,
+                catch_block.basic_block,
+                c_string.as_ptr(),
+            )
+        };
+
+        unsafe {
+            CallSiteValue::new(value)
+        }
+    }
+
+    /// Build a cleanup landing pad. This landing pad is always visited when unwinding the stack.
+    /// A cleanup is extra code that needs to be run when unwinding a scope. C++ destructors are a
+    /// typical example. In a language with reference counting, the cleanup block can decrement the
+    /// refcount of values in scope.
+    ///
+    /// The personality function determines how an exception is handled.
+    /// This example shows how to use the C++ personality function:
+    ///
+    /// ```no_run
+    /// let context = Context::create();
+    /// let module = context.create_module("sum");
+    /// let builder = context.create_builder();
+    ///
+    /// let f32_type = context.f32_type();
```

Comment:
so, should this full example be here? it's big, but also I don't see a better place to put it.

Reply:
This is fine if you think it's helpful